File: python/sorting/merge.py

def merge(left, right):
    result = []

    while left and right:
        if left[0] <= right[0]:
            result.append(left[0])
            left = left[1:]
        else:
            result.append(right[0])
            right = right[1:]

    if left:
        result += left
    if right:
        result += right

    return result


def sort(lst):
    if len(lst) <= 1:
        return lst

    middle = int(len(lst) / 2)
    left, right = sort(lst[:middle]), sort(lst[middle:])

    return merge(left, right)

File: golang/sorting/merge/merge.go
package merge

func Sort(list []int) []int {
	if len(list) <= 1 {
		return list
	}

	middle := len(list) / 2

	left, right := Sort(list[:middle]), Sort(list[middle:])
	return merge(left, right)
}

func merge(left []int, right []int) []int {
	result := make([]int, 0, len(left)+len(right))

	for len(left) > 0 && len(right) > 0 {
		if left[0] <= right[0] {
			result = append(result, left[0])
			left = left[1:]
		} else {
			result = append(result, right[0])
			right = right[1:]
		}
	}

	if len(left) > 0 {
		result = append(result, left...)
	}

	if len(right) > 0 {
		result = append(result, right...)
	}

	return result
}
